### DISCUSSION TITLE
Reality check: How useful exactly are SuiteSparse and LDLFactorizations' LDLT?

### DISCUSSION THREAD

**Discussion by PetrKryslUCSD** [JuliaSmoothOptimizers/LDLFactorizations.jl#103] [General]
Please don't consider this a bashing of a good effort to equip Julia with a general native solver.
It can be useful to some people, no doubt. But for sufficiently large problems these factorizations are hopeless.

Example: A 3D model of a rectangular cantilever block. Roughly 100k equations, 7M non zeros.

```
LDLFactorizations   SuiteSparse ldlt   Skyline ldlt      SuiteSparse cholesky
------------------------------------------------------------------------------
  1414 seconds        1477 seconds       331 seconds          11 seconds (!!!)
```
Two orders of magnitude slow down! 

This is not the case of the implementation being too slow. This is a case of the algorithm not being fast enough.

My point is: we really need to get a native implementation of one of the "supernodal"
factorizations. It is not enough to have access to HSL, MUMPS, or SuiteSparse in dynamically loaded libraries. It leads to duplication of memory, preventing these methods from reaching their full potential in Julia.

**Comment by dpo** [comment]
I would still say "pretty useful" given
1. the lack of alternatives in Julia;
2. the match with SuiteSparse's ldlt's timing in your benchmark;
3. the fact that this packages extends SuiteSparse's ldlt in useful ways (ordering, only one triangle need be stored, and the code is generic).

This package addresses a specific need for us as solving SQD systems is central to our research.

That said, of course, a native multifrontal / supernodal factorization is dearly missing.

Your question over at https://github.com/JuliaSmoothOptimizers/HSL.jl/discussions/85 made me think that it may be allowed to translate SSIDS to Julia (https://github.com/ralna/spral/tree/master/src/ssids). I'll check with the powers that be if that's even legal (it seems to have an MIT-like license). Even if it is, it will probably be a major undertaking.

**Reply by PetrKryslUCSD** [reply]
@dpo I am a bit surprised that you don't mind the long run times of the ldlt (from either package). Are your matrices extremely sparse?

For my applications, the above-mentioned problem of 100K unknowns and 7M nzs is *small*. Imagine the run times of ldlt if applied to 1.5M unknowns, 100M nzs... I haven't tried with the two LDLT packages, but Skyline ran for more than 9 hours.

ssids looks interesting! But C++ :-( There is some f90 though...

**Reply by RoyiAvital** [reply]
I thought that for those sizes iterative solvers are much more recommended than direct ones.

Regarding Windows, I think Intel based `Pardiso.jl` should work and should be competitive.

**Reply by PetrKryslUCSD** [reply]
Yes, iterative solvers are good (when they manage to converge). In my applications, the condition number often causes the iteration to fail.

**Reply by dpo** [reply]
We have quite a number of iterative solvers in Krylov.jl. Which ones did you try? If you have challenging systems, we would be interested in having access to some of them.

**Reply by PetrKryslUCSD** [reply]
I am trying out PCG. ILU(0) usually fails, and ILU(T) tends to be very expensive.

**Reply by RoyiAvital** [reply]
Maybe we should try to improve the incomplete decompositions as well.

I remember seeing a package about random pre conditioners in Julia. I don't recall the name.

**Update**:
OK, I found it: https://github.com/tjdiamandis/RandomizedPreconditioners.jl.

**Reply by PetrKryslUCSD** [reply]
> We have quite a number of iterative solvers in Krylov.jl. Which ones did you try? If you have challenging systems, we would be interested in having access to some of them.

I will set something up. I would be genuinely interested in what you manage to improve. Thanks!

**Reply by dpo** [reply]
> ILU(0) usually fails, and ILU(T) tends to be very expensive.

You might try LimitedLDLFactorizations.jl. At least that will take advantage of the symmetry and SQD-ness of your system. If your matrix is SQD, the preconditioner will be indefinite, so you'll have to use L |D| Lᵀ.

**Reply by PetrKryslUCSD** [reply]
> We have quite a number of iterative solvers in Krylov.jl. Which ones did you try? If you have challenging systems, we would be interested in having access to some of them.

Here are few matrices of various sizes. This is for a nearly incompressible structure.
http://hogwarts.ucsd.edu/~pkrysl/shared/matrices/

The files are HDF5. The package DataDrop can be used to load the matrix as CSC:
```
using DataDrop
K = DataDrop.retrieve_matrix("K63070.h5")
```

**Reply by dpo** [reply]
Thanks. What are the properties of those systems? (symmetric? posdef? ...)

**Reply by amontoison** [reply]
@PetrKryslUCSD 
I tried a `lldl` preconditioner for your problems.
I share the code if you want to test it too.

```julia
using LimitedLDLFactorizations, LinearOperators, Krylov
using LinearAlgebra, SparseArrays, MKLSparse, DataDrop

function ildl(K::AbstractMatrix{T}; kwargs...) where T
    F = lldl(K; kwargs...)
    F.D .= abs.(F.D)
    n = length(F.D)
    return LinearOperator(T, n, n, true, true, (y, v) -> ldiv!(y, F, v))
end

pbs = ["K28782", "K43710", "K63070", "K87438", "K117390"]

atol = 1e-10
rtol = 0.0
memory = 2000

for name in pbs
  A = DataDrop.retrieve_matrix("./pkrysl/$name.h5")
  n, m = size(A)
  b = A * ones(n)

  println("Problem: ", name)
  P = ildl(A, memory=memory)
  # x, stats = minres_qlp(A, b, M=P, atol=atol, rtol=rtol, verbose=1))
  x, stats = bicgstab(A, b, N=P, atol=atol, rtol=rtol, verbose=1)
  println()
end
```

**Reply by PetrKryslUCSD** [reply]
I got CG to converge for 63070 unknowns in 284 iterations. No success for the larger matrices.

**Reply by amontoison** [reply]
@PetrKryslUCSD Can you upload the right-hand side that you have for your linear systems ?
I will try to improve what I proposed.

**Reply by PetrKryslUCSD** [reply]
Hi @amontoison, the right-hand sides are there (`F...`), as are the corresponding matrices (`K...`): http://hogwarts.ucsd.edu/~pkrysl/shared/matrices/

**Comment by dpo** [comment]
I do mind but waiting for those factorizations to finish is many orders of magnitude shorter than writing a multifrontal/multicore/supernodal factorization, which is not my specialty.

**Reply by PetrKryslUCSD** [reply]
I do need an LDLT, btw, because the diagonal elements of factorized Jacobians reveal properties of mechanical systems during Newton iterations. And, they can be both positive and negative.

**Comment by PetrKryslUCSD** [comment]
How are your sparse matrices subject to LDLT structured? If you could live with envelope storage, my package SkylineSolvers would be quite a bit faster than SuiteSparse LDLT... So, should you be interested, I'd be happy to make it available to you.

**Reply by dpo** [reply]
Sure, I would be happy to give it a try! Our systems typically have a saddle-point form with a diagonal bottom block. The other blocks can be quite general.

**Reply by PetrKryslUCSD** [reply]
Cool. I will make sure there is at least some modicum of user friendliness and I will make the package public.

**Reply by PetrKryslUCSD** [reply]
Check out https://github.com/PetrKryslUCSD/SkylineSolvers.jl, and do let me know how things work out please.
Feedback is appreciated.

**Comment by PetrKryslUCSD** [comment]
Symmetric, PD.

**Comment by PetrKryslUCSD** [comment]
@amontoison  That is very interesting, many thanks!

I did not get it to converge for 63070 unknowns.
I killed the process at
```
  551  1.6e+02   5.5e-01   3.4e-02

  552  1.6e+02   7.6e-03  -3.3e-03

  553  1.6e+02  -2.3e-02   7.4e-02

ERROR: InterruptException:
```
I got it to converge for the two smaller matrices (uploaded today to
http://hogwarts.ucsd.edu/~pkrysl/shared/matrices/)
and the corresponding right hand side vectors.

The iteration would not converge for the two larger matrices that I
uploaded today. Not even when I increased memory=4000.
I killed the iterations after about five hundred, when the residual was
showing no signs of decreasing.

I use almost precisely the code that you gave me:
```
        atol = 1e-10
        rtol = 0.0
        memory = 4000
        @time P = ildl(K, memory=memory)
        @time U, stats = bicgstab(K, F, N=P, atol=atol, rtol=rtol,
verbose=1)
        @show stats
```

Petr Krysl
Prof. and Vice chair for undergraduate education
Structural engineering department
University of California, San Diego

**Reply by PetrKryslUCSD** [reply]
Hi Alexis @amontoison,

Have you had any luck with those linear systems?

P

**Reply by amontoison** [reply]
Hi @PetrKryslUCSD,
Sorry I was very busy recently but I will work on it tomorrow !

**Reply by amontoison** [reply]
```julia
using AlgebraicMultigrid, LinearOperators, Krylov
using LinearAlgebra, SparseArrays, MKLSparse, DataDrop, Metis

function amg(A::AbstractMatrix{T}; kwargs...) where T
  n, m = size(A)
  ml = ruge_stuben(A; kwargs...)
  # ml = smoothed_aggregation(A; kwargs...)
  P = aspreconditioner(ml)
  return LinearOperator(T, n, n, true, true, (y, v) -> ldiv!(y, P, v))
end

pbs = ["43710", "101680", "196350", "336720"]

atol = 1e-10
rtol = 0.0

for name in pbs
  println("Problem: ", name)
  A = DataDrop.retrieve_matrix("./K$name.h5")
  b = DataDrop.retrieve_matrix("./F$name.h5")[:]
  P = amg(A)
  x, stats = minres_qlp(A, b, M=P, atol=atol, rtol=rtol, verbose=1)
end
```

**Reply by amontoison** [reply]
@PetrKryslUCSD, I tested [`AlgebraicMultigrid.jl`](https://github.com/JuliaLinearAlgebra/AlgebraicMultigrid.jl) package and it works well for your problems.
AMG preconditioners are often used to solve discretized PDEs but It's not my field so I never tested it before. 
The package is not documented but I found how to build a preconditioner with LinearOperators.jl.
I tested two preconditioners based on the functions `ruge_stuben` (P1) and `smoothed_aggregation` (P2) of `AMG.jl` with the Krylov method `minres_qlp`.

The number of iterations for each problem is:
- ("K43710", P1) -> 1391
- ("K43710", P2) -> 1587
- ("101680", P1) -> 1748
- ("101680", P2) -> 2107 
- ("196350", P1) -> 2105
- ("196350", P2) -> 2535 
- ("336720", P1) -> 2422
- ("336720", P2) -> 2958

**Reply by PetrKryslUCSD** [reply]
That is an interesting observation, thanks! I will check it out.
